### DISCUSSION TITLE
Codec options

### DISCUSSION THREAD

**Discussion by gp-cp** [BeanieODM/beanie#377] [Question]
Hi, 

how to add codec_options to support custom types in mongo? 
Example:
`
from bson.codec_options import TypeCodec, CodecOptions, TypeRegistry

class SetCodec(TypeCodec):

    python_type = set
    bson_type = list

    def transform_python(self, value):
        return list(value)

    def transform_bson(self, value):
        return value

set_codec = SetCodec()
type_registry = TypeRegistry([set_codec])
codec_options = CodecOptions(type_registry=type_registry)
client.get_database(name="foobar", codec_options=codec_options)
`

**Comment by IGJAmpere** [comment]
Hi! 
I am also interested in `pymongo`'s `CodecOptions` for [ `datetime` readings](https://pymongo.readthedocs.io/en/stable/examples/datetimes.html), as my code was breaking because it was not able to compare naive to non-naive `datetime`s. 

I have been able to apply this at the database level changing the system of retrieving/generating the database, like this: 
```python
    from motor import AsyncIOMotorClient
    from bson.codec_options import CodecOptions
    from datetime import timezone

    CODEC_OPTIONS = CodecOptions(tz_aware=True, tzinfo=timezone.utc)
    client = AsyncIOMotorClient("mongodb://localhost:27017")
    db = client.get_database("test_db", codec_options=CODEC_OPTIONS)
    await init_beanie(database=db, document_models=...)
```

But it would be nice to have this `CodecOptions` accessible maybe on a _per collection_ level (as shown in `pymongo`'s docs above), possibly through `Document`'s `Settings` inner class.

Maybe the option is already there, hopefully something like this will be documented.

**Reply by roman-right** [reply]
This sounds reasonable. I'll add this feature to my ToDo list.

**Reply by Mark90** [reply]
Nice, I wasn't aware of the CodecOptions fix!

Since the BSON date type is considered UTC, would it perhaps make sense for beanie to always return datetime objects with tzinfo set to utc?

**Reply by IGJAmpere** [reply]
In Python, naive `datetime` objects are basically the default (sadly, in my opinion), so I think enforcing UTC timezones might cause more unexpected problems to some users, as you cannot operate naive with aware (it crashes). 

So, I'd say the best solution is for this to be easily _toggleable_, and well documented in Beanie.

---

As a side note, these Python guys thought that `datetime.utcnow` had to return a _naive_ `datetime`... so... yeah...

**Reply by Mark90** [reply]
> So, I'd say the best solution is for this to be easily _toggleable_, and well documented in Beanie.

Fully agree, it could be part of an option like `strict_datetimes` which, when enabled, would make documents return (and require) aware datetimes. But I don't know how much extra work this puts on Roman's plate :) 

> As a side note, these Python guys thought that `datetime.utcnow` had to return a _naive_ `datetime`... so... yeah...

Yep, I think this contributes to keeping developers in the dark about timezones.
